daemon/cmd/policy.go

```diff
@@ -452,20 +458,52 @@ func (d *Daemon) policyAdd(sourceRules policyAPI.Rules, opts *policy.AddOptions,
 // to a policy repository for a daemon. This currently consists of endpoint
 // regenerations / policy revision incrementing for a given endpoint.
 type PolicyReactionEvent struct {
-	wg                *sync.WaitGroup
-	epsToBumpRevision *policy.EndpointSet
-	endpointsToRegen  *policy.EndpointSet
-	newRev            uint64
-	upsertIdentities  map[string]*identity.Identity // deferred CIDR identity upserts, if any
-	releasePrefixes   []*net.IPNet                  // deferred CIDR identity deletes, if any
+	wg                      *sync.WaitGroup
+	epsToBumpRevision       *policy.EndpointSet
+	addEndpointsToRegen     *policy.EndpointSet
+	deletedEndpointsToRegen *policy.EndpointSet
+	newRev                  uint64
+	upsertIdentities        map[string]*identity.Identity // deferred CIDR identity upserts, if any
+	releasePrefixes         []*net.IPNet                  // deferred CIDR identity deletes, if any
+	policyOperation         string
 }
 
 // Handle implements pkg/eventqueue/EventHandler interface.
 func (r *PolicyReactionEvent) Handle(res chan interface{}) {
 	// Wait until we have calculated which endpoints need to be selected
 	// across multiple goroutines.
 	r.wg.Wait()
-	reactToRuleUpdates(r.epsToBumpRevision, r.endpointsToRegen, r.newRev, r.upsertIdentities, r.releasePrefixes)
+	operationHandle(r.epsToBumpRevision, r.deletedEndpointsToRegen, r.addEndpointsToRegen, r.newRev, r.upsertIdentities, r.releasePrefixes, r.policyOperation)
+}
+
+func operationHandle(epsToBumpRevision, deletedEndpointsToRegen *policy.EndpointSet, addEndpointsToRegen *policy.EndpointSet, newRev uint64, upsertIdentities map[string]*identity.Identity, releasePrefixes []*net.IPNet, policyOperation string) {
+
+	if policyOperation == "update" {
+		if deletedEndpointsToRegen.Len() != 0 { // Delete in use Policy
+			metrics.PolicyUsed.Dec()
+		}
+		if deletedEndpointsToRegen.Len() == 0 { // delete dummy Policy
+			metrics.PolicyUnused.Dec()
+		}
+		metrics.PolicyUnused.Inc()
+		reactToRuleUpdates(epsToBumpRevision, deletedEndpointsToRegen, newRev, upsertIdentities, releasePrefixes)
+		reactToRuleUpdates(epsToBumpRevision, addEndpointsToRegen, newRev, upsertIdentities, releasePrefixes)
```

Comment:
This would bump policy revision twice on each endpoint, and also upsert Identities twice, as well as release prefixes twice.

I'd much rather see the original `reactToRuleUpdates()` in place with combined endpointset, and have the metrics update be a separate operation to avoid making unnecessary changes to the `reactToRuleUpdates()` calls.